### PR TITLE
Fix support for renorm with float16

### PR DESCRIPTION
**Pull request by Mostafa-Alaa**
This PR fixes a bug when using BatchNormalization with `renorm` set and the input tensor `x` data type is float16 or bfloat16.

A simple code to exploit the bug:
``` python
import tensorflow as tf
from tensorflow.keras.layers import BatchNormalization

X = tf.placeholder(tf.float16, [8, 256, 256, 32], 'X')

renorm_clipping = {
    'rmin': tf.constant(1, tf.float16, [], 'rmin'),
    'rmax': tf.constant(1, tf.float16, [], 'rmax'),
    'dmax': tf.constant(0, tf.float16, [], 'dmax'),
}

BatchNormalization(renorm=True, renorm_clipping=renorm_clipping)(X, True)
```

### CODE REVIEW

**Comment by georgepaw** [comment]
Could we have a look at this please?

**Comment by qlzh727** [comment]
Nudge @fchollet for review.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by qlzh727** [comment]
Thanks for the change and sorry for the long delay. The code looks good, can you add a unit test for that? Thanks.

**Comment by Mostafa-Alaa** [comment]
No problem, I'll be adding the test unit asap.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thank you for the PR.

1) Before casting any tensor, you should do a type check to see if it needs to be cast, e.g.

```python
if x.dtype != target_dtype:
   x = cast(x, target_dtype)
```
This avoids unnecessary cast ops in the graph.

2) Every change should be unit-tested. Please add a unit test that 1) demonstrates the existence of the problem you are fixing (the test should fail without your fix), and 2) verifies the correctness of the new behavior.

Thank you!

**Comment by Mostafa-Alaa** [comment]
@fchollet Thank you for the review.

For the first part I think it might not be the best to add the type check before a cast for a couple of reasons
  1) It would add a lot of redundant code before each cast.
  2) Type checking using ```x.dtype != target_dtype``` would work if ```x``` is a tensor but will fail for python lists, tuples, scalars.

A better design would be to add the type check in the cast function itself which already takes place as per [here](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/ops/math_ops.py#L654) if ```x``` is a tensor, so the unnecessary cast ops wouldn't be a problem, the case where an unnecessary cast op might emerge if ```x``` is not a tensor and a ```convert_to_tensor``` call might be sufficient is already a ```TODO```.

Kindly correct me if I missed something, as for the second part I'll add it right away.

**Comment by Mostafa-Alaa** [comment]
@fchollet @qlzh727 I've added the unit test kindly check and review.

**Comment by qlzh727** [comment]
Ping @fchollet for review.

**Comment by Mostafa-Alaa** [comment]
@qlzh727 I've fixed the indentation problem, please rerun the checks.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by qlzh727** [comment]
@Mostafa-Alaa , can u please resolve the merge conflict? Thanks.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 59 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 74 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 89 days with no activity and the `awaiting review` label has been applied.

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
